Add cluster update callbacks

nt opened the pull request:
This change allows filters to subscribe to Cluster lifecycle events.

Risk Level: Low

junr03 commented:
@nt do you have a specific use case in mind, or are you adding functionality for the potential of it?

mattklein123 commented:
This is for cluster lazy loading. We also need this for allowing Redis to work with CDS.

@nt I can review but this will need tests. Can you please add some to cluster_manager_impl_test?

nt commented:
@mattklein123 I'll add tests now. 

@junr03 this allows to write a filter that blocks a requests until a Cluster definition is present on Envoy.

htuch commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -22,6 +22,26 @@
+
+  /**
+   * onClusterAddOrUpdate is called when a new cluster is added or an existing cluster
+   * is updated in the ClusterManager.
```
Interfaces in `include/envoy` should use Doxygen docs, include `@param`.

nt replied:
Done

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -260,6 +263,7 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     Event::Dispatcher& thread_local_dispatcher_;
     std::unordered_map<std::string, ClusterEntryPtr> thread_local_clusters_;
     std::unordered_map<HostConstSharedPtr, ConnPoolsContainer> host_http_conn_pool_map_;
+    std::unordered_set<Envoy::Upstream::ClusterUpdateCallbacks*> update_callbacks_;
```
I think ordering this might make sense for debugging; otherwise we get non-deterministic callback ordering.

mattklein123 replied:
+1 I would just use a vector probably given low rate of updates.

htuch commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -419,9 +419,12 @@ void ClusterManagerImpl::createOrUpdateThreadLocalCluster(ClusterData& cluster)
-                                                                thread_aware_lb_factory));
+            auto thread_local_cluster = new ThreadLocalClusterManagerImpl::ClusterEntry(
+                cluster_manager, new_cluster, thread_aware_lb_factory);
+            cluster_manager.thread_local_clusters_[new_cluster->name()].reset(thread_local_cluster);
```
I would make this a `make_unique` and then access via `*` below still.

nt replied:
I did, then reverted based on Matt's comment about double deletes.

mattklein123 commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -22,6 +22,26 @@
 namespace Upstream {
 
+/**
+ * ClusterUpdateCallbacks provide a way to exposes Cluster lifecycle event in the
```
nit: s/event/events

nt replied:
Done.

mattklein123 commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -147,6 +167,13 @@ class ClusterManager {
   virtual const std::string& localClusterName() const PURE;
+
+  /**
+   * These two methods allow to register and un register callback for cluster lifecycle events
```
nit: "unregister" and "callbacks"

mattklein123 replied:
Also, can you flesh out the docs here about threading? I see from the code that this effectively needs to be called on every TLS thread to be useful. If we keep this method, I would probably rename to `addThreadLocalClusterUpdateCallbacks` etc. Though this is making me wonder if this is the most useful programming model. I wonder if it would be better for the callbacks to be registered on the main thread once, and then fired on all TLS workers? I suspect that might be easier to use in the general case.

htuch replied:
If you're a filter that is dynamically registering for the duration of a request, you only want the thread-local callback.

mattklein123 replied:
I wasn't suggesting not having a thread local callback, but having to register for it on each worker seems odd. In general, a filter will register for callbacks on the main thread, and then expect to receive them on each worker. That's what I'm suggesting.

htuch replied:
The flow for this filter is that a request comes in for cluster X. Filter registers to hear about clusters while waiting for X to arrive and the unregisters after (it no longer cares about cluster updates). Are you suggesting that it would be preferable to always be listening for cluster updates?

mattklein123 replied:
I see. There are use cases in which I think we do want to always listen and we will want to register on the main thread for updates to fire on each worker. We can start with this, but I would still clarify in the docs.

nt replied:
I updated the docs.

mattklein123 commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -419,9 +419,12 @@ void ClusterManagerImpl::createOrUpdateThreadLocalCluster(ClusterData& cluster)
-                                                                thread_aware_lb_factory));
+            auto thread_local_cluster = std::make_unique<ThreadLocalClusterManagerImpl::ClusterEntry>(
+                cluster_manager, new_cluster, thread_aware_lb_factory);
+            cluster_manager.thread_local_clusters_[new_cluster->name()].reset(thread_local_cluster.get());
```
This code will double delete. You need to call `release()` here. The other way is to just store the allocation in a reference and then move it.

nt replied:
Thank you for catching this; I was struggling to figure out why the integration tests were crashing.

zuercher commented:
@nt fixing the DCO will require re-writing these commits with the exact user.name and user.email your git is configured to use. I think it's fine if you wait until the reviews are finished to preserve the comment history.

nt commented:
Oops. @htuch I rebased and squashed all commits into one and that seems to have "dismissed" your review.

Is normal practice to leave all the commit history? Is that squashed at merge time?

mattklein123 commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -147,6 +170,17 @@ class ClusterManager {
+   *
+   * @param callbacks are the ClusterUpdateCallbacks to add or remove to the cluster manager.
+   */
+  virtual void addClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
```
I still would recommend naming these `addThreadLocal...` and `removeThreadLocal...`. Either way, I would clarify in the method docs that the caller needs to call these functions on every thread in which they expect to receive callbacks. It's not completely clear IMO that is what happens.

nt replied:
Done.

mattklein123 commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -260,6 +263,7 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     Event::Dispatcher& thread_local_dispatcher_;
     std::unordered_map<std::string, ClusterEntryPtr> thread_local_clusters_;
     std::unordered_map<HostConstSharedPtr, ConnPoolsContainer> host_http_conn_pool_map_;
+    std::set<Envoy::Upstream::ClusterUpdateCallbacks*> update_callbacks_;
```
unordered_set?

nt replied:
That is how I had it initially but I was decided that an ordered set would make debugging easier.

mattklein123 replied:
Here is what I would personally do here which will solve this comment as well as my other comment about RAII:

1) Use an `std::list` which gives O(1) add. Coupled with https://github.com/envoyproxy/envoy/blob/master/source/common/common/linked_object.h you can also get O(1) removal.
2) Instead of having an explicit "remove" API, return an opaque handle, which itself just contains the linked entry.
3) When the opaque handle is destroyed, the callback is unlinked. 

I think this is the optimal solution and pretty easy to implement. WDYT?

nt replied:
Done.

mattklein123 commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -147,6 +170,17 @@ class ClusterManager {
+   * @param callbacks are the ClusterUpdateCallbacks to add or remove to the cluster manager.
+   */
+  virtual void addClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
+  virtual void removeClusterUpdateCallbacks(ClusterUpdateCallbacks& callbacks) PURE;
```
One quick question on API. Given that we expect to potentially add/remove these frequently, and will have to deal with filter lifecycle events, should we have an RAII pattern here where `add` returns a handle that implicitly does a remove when it is destroyed?

nt replied:
I am not versed enough in cpp so I'll let others comment on this one.

htuch commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -22,6 +22,38 @@
+public:
+  virtual ~CallbackRegistration() {}
+};
+
```
What about `source/common/common/callback_impl.h`? Could you use this here? If not, I suggest calling this `ClusterUpdateCallbackHandle` and also doing a type alias for the  `std::unique_ptr` variant to `ClusterUpdateCallbackHandlePtr`.

nt replied:
Done: renamed to ClusterUpdateCallbacksHandle and created ClusterUpdateCallbacksHandlePtr.

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -284,6 +288,15 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
   };
 
+  struct CallbackRegistrationImpl : public CallbackRegistration {
+    CallbackRegistrationImpl(ClusterUpdateCallbacks* cb, std::list<ClusterUpdateCallbacks*>& ll);
```
Maybe call `ll` instead `parent`.

nt replied:
Done.

htuch commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -284,6 +288,15 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
   };
 
+  struct CallbackRegistrationImpl : public CallbackRegistration {
+    CallbackRegistrationImpl(ClusterUpdateCallbacks* cb, std::list<ClusterUpdateCallbacks*>& ll);
```
Probably want a `const&` to `cb`.

zuercher commented:
The mac build failed due to a Circle CI's infrastructure, so going ahead with the merge.

nt commented:
Thank you for the patient and thorough review.
